File: internal/diagram/diagram.go
// Пакет diagram — геометричний результат розкладки: фігури з координатами і
// ребра-стрілки. Це «контракт» між layout (виробляє) і render (малює). Пуста
// структура даних, без логіки й залежностей.
package diagram

// Kind — тип фігури ДСТУ.
type Kind int

const (
	Terminator Kind = iota // початок/кінець — овал (стадіон)
	Process                // дія — прямокутник
	Decision               // умова — ромб
	InOut                  // ввід/вивід — паралелограм
)

// Point — точка у координатах діаграми (y росте вниз).
type Point struct{ X, Y float64 }

// Shape — фігура. X,Y — лівий верхній кут; текст центрується.
type Shape struct {
	Kind Kind
	X, Y float64
	W, H float64
	Text string
}

// Edge — стрілка як ортогональна ламана (масив точок) з опційною позначкою.
// Arrowless — без вістря (для з'єднань у точці злиття гілок).
type Edge struct {
	Points    []Point
	Label     string
	Arrowless bool
}

// Diagram — повна розкладена схема.
type Diagram struct {
	Shapes []Shape
	Edges  []Edge
	W, H   float64
}

File: packages/engine/src/diagram.ts
// Модель геометрії — порт pkg/diagram. Контракт між layout (виробляє) і render
// (малює). Чиста структура даних. Kind серіалізується рядком (як у Go MarshalJSON).

export type Kind =
  | 'terminator' // початок/кінець — овал
  | 'process'    // дія — прямокутник
  | 'decision'   // умова — ромб
  | 'io'         // ввід/вивід — паралелограм
  | 'loop'       // початок циклу (for) — шестикутник
  | 'subprogram' // виклик підпрограми — прямокутник з боковими рисками
  | 'connector'; // з'єднувач «А» — коло

export interface Point { x: number; y: number; }

export interface Shape {
  kind: Kind;
  x: number; y: number; w: number; h: number;
  text: string;
}

export interface Edge {
  points: Point[];
  label?: string;
  arrowless?: boolean;
}

export interface Diagram {
  shapes: Shape[];
  edges: Edge[];
  w: number;
  h: number;
  caption?: string;
  figNum?: number;
  capWord?: string;
  capFormat?: string;
}

export const CAPTION_WORD = 'Рисунок';
export const CAP_FORMAT_DEFAULT = '{word} {num} — {text}';

// labelAnchor — позиція й вирівнювання підпису ребра (Так/Ні) за першим сегментом.
export function labelAnchor(p0: Point, p1: Point): { x: number; y: number; align: string } {
  if (p1.y === p0.y && p1.x !== p0.x) { // горизонтальний сегмент
    let off = (p1.x - p0.x) / 2;
    if (off > 34) off = 34;
    else if (off < -34) off = -34;
    return { x: p0.x + off, y: p0.y - 9, align: 'middle' };
  }
  if (p1.x === p0.x) { // вертикальний сегмент
    return { x: p0.x + 10, y: (p0.y + p1.y) / 2, align: 'start' };
  }
  return { x: p0.x + 8, y: p0.y - 8, align: 'start' };
}

export function capSupplement(d: Diagram): string { return d.capWord || CAPTION_WORD; }
function capFmt(d: Diagram): string { return d.capFormat || CAP_FORMAT_DEFAULT; }

// capSeparator — роздільник між номером і текстом (для нативного підпису Typst-figure).
export function capSeparator(d: Diagram): string {
  const fmt = capFmt(d);
  const i = fmt.indexOf('{num}');
  const j = fmt.indexOf('{text}');
  if (i < 0 || j < 0 || j < i) return ' — ';
  return fmt.slice(i + '{num}'.length, j);
}

// capHasWord — чи містить шаблон слово-supplement (для Typst-figure).
export function capHasWord(d: Diagram): boolean { return capFmt(d).includes('{word}'); }

// captionLine — повний рядок підпису («Рисунок N — текст») за шаблоном.
export function captionLine(d: Diagram): string {
  if (!d.caption) return '';
  if (!d.figNum || d.figNum <= 0) return d.caption;
  return capFmt(d)
    .replaceAll('{word}', capSupplement(d))
    .replaceAll('{num}', String(d.figNum))
    .replaceAll('{text}', d.caption)
    .trim();
}
